python/cudf/cudf/core/multiindex.py

```diff
@@ -114,7 +114,11 @@ def __init__(
                 )
             else:
                 level = DataFrame({name: self._levels[i]})
-            level = DataFrame(index=codes).join(level)
+
+            level = DataFrame(
+                {"original_order": self.codes[name].index}, index=codes
+            ).join(level)
+            level = level.sort_values(by="original_order").reset_index(True)
             source_data[name] = level[name].reset_index(drop=True)
```

Comment:
In addition to this fix, should we just update the `from_product` constructor to call `.to_frame()` against the Pandas MultiIndex so we can avoid this heavier construction path anyway?

Reply:
With the change introduced in this commit : https://github.com/rapidsai/cudf/pull/4588/commits/d47c60aeb6afdc2f0e41dcc213adab4b4dd0431c

Time taken by the tests(`test_multiindex_from_product`) to run:
**4.47s**

Before this change:
**19.10s**